finagle/buoyant/src/main/scala/com/twitter/finagle/buoyant/PathMatcher.scala

```diff
@@ -58,33 +58,47 @@ object PathMatcher {
 
   private class Matcher(expr: String) extends PathMatcher {
 
-    val exprSegments = expr.split("/").dropWhile(_.isEmpty)
-
     override def extract(path: Path): Option[Map[String, String]] =
-      _extract(path.showElems, exprSegments, Map.empty)
+      _extract(path.showElems, regexSegments, Map.empty)
+
+    private[this] val segmentRegex = """\{([a-zA-Z0-9\.:-]+)\}""".r
+
+    private[this] def mkRegex(x: String, y: String) = x.replace(y, "(.*)")
+
+    private[this] val regexSegments: Seq[(Option[scala.util.matching.Regex], Seq[String])] = {
```

Comment:
What do each part of this return type represent?  As written, I think the second part represents the list of capture keys when the segment contains captures but it represents the literal segment when there are no captures.  This reassignment of meaning is confusing.  To make this explicit, consider a case class:

```
case class MatchSegment(
  segment: String,  
  regex: Option[Regex],
  captureKeys: Seq[String]
)
```

or something like

```
sealed trait MatchSegment
case object Wildcard extends MatchSegment
case class Literal(segment: String) extends MatchSegment
case class Captures(regex: Regex, keys: Seq[String]) extends MatchSegment
```